### PR TITLE
Test resources

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

**Comment by sdelamo** [comment]
@melix  and @alvarosanchez draft Pr for test resources

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataJdbcSpec.groovy [hunk 1]
```diff
@@ -91,48 +92,41 @@ class DataJdbcSpec extends ApplicationContextSpec  implements CommandOutputFixtu
...
         when:
-        ctx = buildGeneratorContext(['data-jdbc', 'sqlserver'])
+        GeneratorContext ctx = buildGeneratorContext(features, options)
 
         then:
-        ctx.configuration.containsKey("datasources.default.url")
-        ctx.configuration.get("datasources.default.dialect") == "SQL_SERVER"
+        if (buildTool == BuildTool.MAVEN) {
+            assert ctx.configuration.containsKey("datasources.default.url")
+        }
```
Comment (timyates) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataJdbcSpec.groovy]: we should check Gradle doesn't contain this key

File: starter-core/src/main/java/io/micronaut/starter/feature/database/Oracle.java [hunk 2]
```diff
@@ -22,17 +22,20 @@
 
 import io.micronaut.starter.feature.database.r2dbc.R2dbc;
 import io.micronaut.starter.feature.migration.MigrationFeature;
+import io.micronaut.starter.feature.testresources.TestResources;
 import jakarta.inject.Singleton;
 
 @Singleton
 public class Oracle extends DatabaseDriverFeature {
 
     public static final String NAME = "oracle";
```
Comment (timyates) [inline starter-core/src/main/java/io/micronaut/starter/feature/database/Oracle.java]: Should we use `NAME` in the other DB features?

We can then use Oracle.NAME in the tests

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataMongoSpec.groovy [hunk 3]
```diff
@@ -24,13 +24,15 @@ class DataMongoSpec extends ApplicationContextSpec implements CommandOutputFixtu
...
-        } else {
+        } else if (buildTool == BuildTool.MAVEN) {
             assert build.contains('testImplementation("org.testcontainers:junit-jupiter")')
         }
-        build.contains('testImplementation("org.testcontainers:mongodb")')
-        build.contains('testImplementation("org.testcontainers:testcontainers")')
+        if (buildTool == BuildTool.MAVEN) {
+            assert build.contains('testImplementation("org.testcontainers:mongodb")')
+            assert build.contains('testImplementation("org.testcontainers:testcontainers")')
+        }
```
Comment (timyates) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataMongoSpec.groovy]: These MAVEN tests seem to be checking for Gradle dependencies?

**Review by timyates — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
@timyates could you merge master into this and continue?

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataMongoSpec.groovy (hunk 3 above)
Comment (alvarosanchez) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataMongoSpec.groovy]: Fixed

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataJdbcSpec.groovy (hunk 1 above)
Comment (alvarosanchez) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataJdbcSpec.groovy]: done

**Comment by timyates** [comment]
@alvarosanchez There's another one here I think: https://github.com/micronaut-projects/micronaut-starter/pull/1297/files#diff-06e8e0b1f3f6b577854a81f5d5791ab09e7cbc5ba0cf8720f394ce45526e8dc5R168

File: starter-core/src/main/resources/pom.xml [hunk 4]
```diff
@@ -113,6 +113,11 @@
             <artifactId>micronaut-gradle-plugin</artifactId>
             <version>3.4.1</version>
         </dependency>
+        <dependency>
+            <groupId>io.micronaut.gradle</groupId>
+            <artifactId>micronaut-test-resources-plugin</artifactId>
+            <version>3.5.0-SNAPSHOT</version>
```
Comment (melix) [inline starter-core/src/main/resources/pom.xml]: Can use 3.5.0 final now

**Comment by alvarosanchez** [comment]
@melix `CreateMessagingSpec` is failing only for the feature `jms-oracle-aq`. For Gradle, the error is:

```
Message: Error configuring data source 'default'. No URL specified
```

No idea if that should work or not. All the other messaging features seem to be working.

In the case of Maven, they fail because of this was missing: https://github.com/micronaut-projects/micronaut-core/pull/7676

**Comment by timyates** [comment]
Are messaging features using test resources here yet?

I thought we had to do something to get RabbitMQ and Kafka using test resources

And then add something to make these use a "shared" server

**Comment by melix** [comment]
Test resources currently support Kafka, MQTT and RabbitMQ. I'm trying to find my way in how all those interact in starter.

**Comment by melix** [comment]
There are actually quite a few failures related to DB features. I think one of the issues is that test resources inference works with Micronaut Data, but not Micronaut SQL.

**Comment by timyates** [comment]
I'll take a look, got a scan?

**Comment by timyates** [comment]
Or is this just maven?

**Comment by timyates** [comment]
@sdelamo This should now define the database module in the plugin config (when we are using JDBC without Data), and add the `<shared>` flag when we are using Rabbit, Mqtt5, Kafka or KafkaStream

**Comment by timyates** [comment]
Whoops, @alvarosanchez not Sergio ☝️  #muscle-memory

**Comment by timyates** [comment]
@alvarosanchez There seems to be an issue with Maven and messaging features:

https://ge.micronaut.io/s/ejrkbnqod2i56/tests/overview?outcome=failed

Do you know what

```
[ERROR]   The project example.micronaut:foo:0.1 (/tmp/agorapulse-micronaut-permissions9117599170926901999/pom.xml) has 1 error |  
[ERROR]     'dependencies.dependency.version' for io.micronaut.testresources:micronaut-test-resources-client:jar is missing. @ line 134, column 17
```

means?

**Comment by timyates** [comment]
Must be something I missed from "ApplicationType.MESSAGING"

**Comment by timyates** [comment]
I wonder if it's because they are all MESSAGING apps, and they don't seem to have

```
    <micronaut.test.resources.enabled>true</micronaut.test.resources.enabled>
```

in their pom properties

**Comment by alvarosanchez** [comment]
No, it's because this was missing: https://github.com/micronaut-projects/micronaut-core/pull/7676

However, core's Java CI is broken: https://github.com/micronaut-projects/micronaut-core/actions/workflows/gradle.yml?query=branch%3A3.6.x, hence no new snapshot was published.

@melix worked on a fix (https://github.com/micronaut-projects/micronaut-build/pull/361), but I guess we need a new release?

**Comment by timyates** [comment]
@alvarosanchez Basically the same error in test-core https://ge.micronaut.io/s/bw7qukcsmret4/tests/overview?outcome=failed

**Comment by alvarosanchez** [comment]
As a workaround to move forward, you can (temporarily) make the feature to add the following property: `micronaut.test.resources.version=1.0.0`

**Comment by melix** [comment]
> @melix worked on a fix (https://github.com/micronaut-projects/micronaut-build/pull/361), but I guess we need a new release?

Yes, I need to release a new build plugins version, but I'll do that on EU night time to avoid creating too many PRs during work time. In the meantime you can disable binary checks on the `inject-java-test` module by adding this:

```
micronautBuild {
    binaryCompatibility {
        enabled = false
    }
}
```

but it looks like folks tend to forget to re-enable, so...

File: starter-core/src/main/java/io/micronaut/starter/feature/testresources/TestResources.java [hunk 5]
```diff
@@ -67,10 +67,15 @@ public void apply(GeneratorContext generatorContext) {
...
 
-            generatorContext.addDependency(Dependency.builder()
+            Dependency.Builder clientDependency = Dependency.builder()
                     .groupId("io.micronaut.testresources")
                     .artifactId("micronaut-test-resources-client")
-                    .test());
+                    .test();
+
+            // TODO: Remove when it's in the BOM
+            clientDependency = clientDependency.version("1.0.0-M2");
```
Comment (melix) [inline starter-core/src/main/java/io/micronaut/starter/feature/testresources/TestResources.java]: Use `1.0.0`, not a milestone.

Comment (timyates) [reply starter-core/src/main/java/io/micronaut/starter/feature/testresources/TestResources.java]: Ooopsie

**Comment by timyates** [comment]
That should be it...  15th time's the charm ☘️

**Comment by timyates** [comment]
@alvarosanchez did you mean to close this 😱

**Comment by alvarosanchez** [comment]
No way! I think the Maven PR was mistakenly linked to this one

**Comment by timyates** [comment]
@melix I think the Ms-sql tests are failing as we have to accept some sort of license to get the container running?

https://ge.micronaut.io/s/cgt2ytn6hjasi/tests/overview?outcome=failed

```
Caused by: java.lang.IllegalStateException: The image mcr.microsoft.com/mssql/server:2017-CU12 requires you to accept a license agreement. Please place a file at the root of the classpath named container-license-acceptance.txt, e.g. at src/test/resources/container-license-acceptance.txt. This file should contain the line: |  
-- | --
  | mcr.microsoft.com/mssql/server:2017-CU12 |  
  | at org.testcontainers.utility.LicenseAcceptanc
```

**Comment by timyates** [comment]
I guess there's a question of whether we:

1) Automatically accept the license for people 😟
2) Document they have to do this, and fix the tests to do it

**Comment by melix** [comment]
That's correct, the license needs to be added to the server classpath. With Gradle that's fairly straightforward, since you can use `src/testResources/resources`. For Maven I'm not sure how to do this :thinking:

**Comment by timyates** [comment]
I'll try and fix the tests 👍

**Comment by timyates** [comment]
Ahhh found the issue...

We *do* write out the EULA here

https://github.com/micronaut-projects/micronaut-starter/blob/f7902691bb355c8a4f923dce086b113c846b8346/starter-core/src/main/java/io/micronaut/starter/feature/database/TestContainers.java#L64

But we write out `mcr.microsoft.com/mssql/server:2019-CU4-ubuntu-16.04`

Whereas TestResources uses `2017-CU12`

https://github.com/micronaut-projects/micronaut-test-resources/blob/b58eb04d202bb34a2ca5db2b6e1cc90d4f5c8f59/test-resources-jdbc/test-resources-jdbc-mssql/src/main/java/io/micronaut/testresources/mssql/MSSQLTestResourceProvider.java#L36

@melix Is there a reason we use an older version in Test Resources?  I'm not a fan of the `-ubuntu-16.04` bit but there doesn't seem to be a better option when I [look at the list of them](https://mcr.microsoft.com/v2/mssql/server/tags/list) 😞

**Comment by melix** [comment]
Note: Even if you write the EULA, this must appear in the test resources server classpath, not the application classpath.

There's no particular reason why we use an older version, except that's the one I found in the test containers docs.

**Comment by timyates** [comment]
> Note: Even if you write the EULA, this must appear in the test resources server classpath, not the application classpath.

~~How do we do that 🤔~~

Found it (for Gradle), `src/testResources/resources/container-license-acceptance.txt`

> There's no particular reason why we use an older version, except that's the one I found in the test containers docs.

I guess we should align them 🤔

**Comment by melix** [comment]
> I guess we should align them thinking

Don't hesitate to create a PR on test resources.

**Comment by timyates** [comment]
@sdelamo If someone specifically asks for TestContainers, should we stop including TestResources?

@alvarosanchez How do we get a file onto the TestResources classpath (EULA for SqlServer) with Maven?

**Comment by alvarosanchez** [comment]
> @alvarosanchez How do we get a file onto the TestResources classpath (EULA for SqlServer) with Maven?

That's not supported, unfortunately. We would need to agree to a conventional directory (which could be `src/testResources/resources`) which our Maven plugin will then make it available to the Test Resources service.

**Comment by timyates** [comment]
> > @alvarosanchez How do we get a file onto the TestResources classpath (EULA for SqlServer) with Maven?
> 
> That's not supported, unfortunately. We would need to agree to a conventional directory (which could be `src/testResources/resources`) which our Maven plugin will then make it available to the Test Resources service.

@melix solved the issue by adding a configuration to accept the EULA 🎉

**Comment by timyates** [comment]
@alvarosanchez I removed both the maven related TODOs from here on Monday (just incase you had a post-it)

**Comment by timyates** [comment]
This still needs Hibernate Reactive PLUS Migration at the same time thinking about...

I think the easiest way is just to do it all in here...instead of [over here](https://github.com/micronaut-projects/micronaut-starter/pull/1323)

**Comment by timyates** [comment]
Think the failure is waiting for a newer maven plugin that picks up the latest test-resources

**Comment by alvarosanchez** [comment]
Note that test-resources version isn't set in the Maven plugin, but in the Micronaut BOM.

I have bumped manually the version here: https://github.com/micronaut-projects/micronaut-core/pull/7739

**Comment by alvarosanchez** [comment]
@timyates 
image

**Comment by timyates** [comment]
@sdelamo Please take a look.  I think all we are lacking is hibernate reactive support, but that needs a 3.6.0 release to be added to test-resources

Actually for now, I guess we should be writing the JDBC url and credentials for Hibernate Reactive support.  I will check we are currently doing that....

**Comment by timyates** [comment]
Hibernate Reactive looks good

https://github.com/micronaut-projects/micronaut-starter/blob/aafe276d9f4175de08a0d6cbaa06e8577c4c88c0/starter-core/src/test/groovy/io/micronaut/starter/feature/database/DataHibernateReactiveSpec.groovy#L257-L262

AFAICS, we're go for launch 🚀

**Pull request merged**